src/aspire/volume/__init__.py

```diff
@@ -287,6 +358,91 @@ class FBBasisVolume(BasisVolume):
     pass
 
 
+def gaussian_blob_Cn_vols(
+    L=8, C=2, K=16, alpha=1, subscript=1, seed=None, dtype=np.float64
+):
+    """
+    Generate Cn rotationally symmetric volumes composed of Gaussian blobs.
+    The volumes are symmetric about the z-axis.
+
+    Defaults to volumes with no symmetry.
+
+    :param L: The size of the volumes
+    :param C: The number of volumes to generate
+    :param K: The number of blobs each volume is composed of.
+    A Cn symmetric volume will be composed of n times K blobs.
+    :param subscript: The subcript n indicating Cn symmetry.
+    :param alpha: A scale factor of the blob widths
+
+    :return: A Volume instance containing C Gaussian blob volumes with Cn symmetry.
+    """
+
+    def _eval_gaussian_blobs(L, Q, D, mu, subscript, dtype=np.float64):
+        try:
+            subscript = int(subscript)
+        except Exception:
+            raise NotImplementedError(
+                f"C{subscript} symmetry not supported. Only Cn symmetry, where n is an integer, is supported."
+            )
+
+        g = grid_3d(L, dtype=dtype)
+        coords = np.array(
+            [g["x"].flatten(), g["y"].flatten(), g["z"].flatten()], dtype=dtype
+        )
+
+        K = Q.shape[-1]
+        vol = np.zeros(shape=(1, coords.shape[-1])).astype(dtype)
+        rot = np.zeros(shape=(3, 3, subscript)).astype(dtype)
+
+        for k in range(subscript):
+            rot[:, :, k] = [
+                [
+                    np.cos(2 * np.pi * k / subscript),
+                    -np.sin(2 * np.pi * k / subscript),
+                    0,
+                ],
+                [
+                    np.sin(2 * np.pi * k / subscript),
+                    np.cos(2 * np.pi * k / subscript),
+                    0,
+                ],
+                [0, 0, 1],
+            ]
+
+        for k in range(K):
+            for j in range(subscript):
+                coords_k = rot[:, :, j] @ coords - mu[:, k, np.newaxis]
+                coords_k = (
+                    Q[:, :, k] / np.sqrt(np.diag(D[:, :, k])) @ Q[:, :, k].T @ coords_k
+                )
+
+                vol += np.exp(-0.5 * np.sum(np.abs(coords_k) ** 2, axis=0))
+
+        vol = np.reshape(vol, g["x"].shape)
+
+        return vol
+
+    def gaussian_blobs(K, alpha):
+        Q = np.zeros(shape=(3, 3, K)).astype(dtype)
+        D = np.zeros(shape=(3, 3, K)).astype(dtype)
+        mu = np.zeros(shape=(3, K)).astype(dtype)
+
+        for k in range(K):
+            V = randn(3, 3).astype(dtype) / np.sqrt(3)
+            Q[:, :, k] = qr(V)[0]
+            D[:, :, k] = alpha ** 2 / 16 * np.diag(np.sum(abs(V) ** 2, axis=0))
+            mu[:, k] = 0.5 * randn(3) / np.sqrt(3)
+
+        return Q, D, mu
+
+    vols = np.zeros(shape=(C, L, L, L)).astype(dtype)
+    with Random(seed):
+        for k in range(C):
+            Q, D, mu = gaussian_blobs(K, alpha)
+            vols[k] = _eval_gaussian_blobs(L, Q, D, mu, subscript, dtype=dtype)
```

Comment:
Since the evaluation happens in a loop with the symmetrization, I don't see a nice way to separate these without storing a fairly large array of rotated coordinates to be accessed by `_eval_gaussians`.

Reply:
Is that array prohibitively large?

Reply:
coords_k would be large, something like (C * L**3 elements) ... assuming you need to store it to evaluate the gaussians in a later loop...  Within the current loop it is consumed right away during the evaluation so we just have a single tmp L**3.

I think it is possible to avoid that, but it is more refactoring (not sure if gaining much clarity there for the effort).  Maybe I'm just not seeing a simpler refactor at this moment.  Its possible to construct a volume and rotate it as well, but I would guess rotating points exactly before eval is less error.

Reply:
@janden Thanks for your help with formulating this. I was able to separate `eval_gaussian_blobs` into two private functions. It may need some tightening up, so I'll revisit it after the weekend.